File: src/backend/bm.js
function bmMatch(text, pattern){
    var last = buildLast(pattern);
    var n = text.length;
    var m = pattern.length;
    var i = m - 1;

    if (i > n - 1){
        return -1;
    }
    var j = m - 1;
    do{
        if(pattern[j] == text[i]){
            if(j==0){
                return i;
            }else{
                i--;
                j--;
            }
        }else{
            lastOcc = last[text[i]];
            i = i + m - Math.min(j, 1+lastOcc);
            j = m-1;
        }
    }while(i <= n-1);
    return false;
}

function buildLast(pattern){
    var last = new Array(128);
    for (var i = 0; i < 128; i++){
        last[i] = -1;
    }
    for (var i = 0; i < pattern.length; i++){
        last[pattern[i]] = i;
    }
    return last;
}


function main(){
    var dna = "ATGCGATACGCTTACGCTTCGAGGAA";
    var disease = "ATGCGATACGTTA";
    console.log(bmMatch(dna, disease));

}
main();

File: src/backend/bm.go
package main

import (
    "fmt"
)

func bmMatch(text string, pattern string) int {
    last := buildLast(pattern)
    n := len(text)
    m := len(pattern)
    i := m - 1

    if i > n-1 {
        return -1
    }

    j := m - 1
    for i <= n-1 {
        if pattern[j] == text[i] {
            if j == 0 {
                return i
            } else {
                i--
                j--
            }
        } else {
            lastOcc := last[text[i]]
            i = i + m - min(j, 1+lastOcc)
            j = m - 1
        }
    }
    return -1
}

func buildLast(pattern string) map[byte]int {
    last := make(map[byte]int)
    for i := 0; i < len(pattern); i++ {
        last[pattern[i]] = i
    }
    return last
}

func min(x, y int) int {
    if x < y {
        return x
    }
    return y
}
